File: packages/sdk-ts/src/admin/tenants.ts
// suite.admin.tenants.* — tenant CRUD.
//
// Endpoint paths and JSON shapes are the canonical contract from
// `apps/dashboard/src/lib/api.ts` (Phase 6 Tenancy section). The wire
// format is snake_case; the SDK exposes camelCase via the helpers in
// `_http.ts`.
//
// All endpoints are gated by the runtime's `multi-tenancy` module flag —
// calls against a disabled runtime surface as `SuiteError` with
// `code === "MT_DISABLED"`.

import { request, toCamel, toSnake, type HttpOptions } from "../_http.js"
import {
  TenantListSchema,
  TenantSchema,
  type CreateTenantInput,
  type Tenant,
  type TenantDetail,
  type TenantList,
  type UpdateTenantInput,
} from "./_models.js"
import { TenantDetailSchema } from "./_models.js"

/** List all non-soft-deleted tenants. */
export async function list(opts: HttpOptions = {}): Promise<TenantList> {
  const raw = await request<unknown>("GET", "/admin/tenants", null, opts)
  return TenantListSchema.parse(toCamel(raw))
}

/** Fetch a tenant's joined detail view (tenant + members + keys + usage). */
export async function get(
  id: string,
  opts: HttpOptions = {},
): Promise<TenantDetail> {
  if (typeof id !== "string" || id.length === 0) {
    throw new Error("tenant id must be a non-empty string")
  }
  const raw = await request<unknown>(
    "GET",
    `/admin/tenants/${encodeURIComponent(id)}`,
    null,
    opts,
  )
  return TenantDetailSchema.parse(toCamel(raw))
}

/** Create a tenant. */
export async function create(
  input: CreateTenantInput,
  opts: HttpOptions = {},
): Promise<Tenant> {
  if (typeof input.slug !== "string" || input.slug.length === 0) {
    throw new Error("input.slug must be a non-empty string")
  }
  if (typeof input.name !== "string" || input.name.length === 0) {
    throw new Error("input.name must be a non-empty string")
  }
  const raw = await request<unknown>(
    "POST",
    "/admin/tenants",
    toSnake(input),
    opts,
  )
  return TenantSchema.parse(toCamel(raw))
}

/** Update selected tenant fields. */
export async function update(
  id: string,
  input: UpdateTenantInput,
  opts: HttpOptions = {},
): Promise<Tenant> {
  if (typeof id !== "string" || id.length === 0) {
    throw new Error("tenant id must be a non-empty string")
  }
  const raw = await request<unknown>(
    "PATCH",
    `/admin/tenants/${encodeURIComponent(id)}`,
    toSnake(input),
    opts,
  )
  return TenantSchema.parse(toCamel(raw))
}

/** Soft-delete a tenant. Returns `true` on success. */
async function deleteTenant(
  id: string,
  opts: HttpOptions = {},
): Promise<boolean> {
  if (typeof id !== "string" || id.length === 0) {
    throw new Error("tenant id must be a non-empty string")
  }
  const raw = await request<{ deleted?: boolean } | null>(
    "DELETE",
    `/admin/tenants/${encodeURIComponent(id)}`,
    null,
    opts,
  )
  if (raw === null || raw === undefined) return true
  return Boolean(raw.deleted ?? true)
}
// `delete` is a reserved word in some import positions, so we re-export
// under both names for ergonomics.
export { deleteTenant as delete }

/** Namespace object — the shape `suite.admin.tenants` is built from. */
export const tenants = {
  list,
  get,
  create,
  update,
  delete: deleteTenant,
} as const

File: packages/sdk-py/af_stack/admin/tenants.py
# SPDX-License-Identifier: Apache-2.0

"""``suite.admin.tenants.*`` — tenant CRUD.

Maps to the admin tenant endpoints in ``apps/dashboard/src/lib/api.ts``:

==================================  =========================================
SDK call                            Endpoint
==================================  =========================================
:func:`list`                        ``GET    /api/v1/admin/tenants``
:func:`get`                         ``GET    /api/v1/admin/tenants/{id}``
:func:`create`                      ``POST   /api/v1/admin/tenants``
:func:`update`                      ``PATCH  /api/v1/admin/tenants/{id}``
:func:`delete`                      ``DELETE /api/v1/admin/tenants/{id}``
==================================  =========================================

All endpoints are gated by the runtime's ``multi-tenancy`` module flag —
calls against a disabled runtime surface as
:class:`af_stack._http.AFStackError` with ``code == "MT_DISABLED"``.
"""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from .. import _http
from ._models import (
    CreateTenantInput,
    Tenant,
    TenantDetail,
    TenantList,
    UpdateTenantInput,
)


async def list() -> TenantList:  # noqa: A001 — mirrors the JS ``tenants.list()`` ergonomics
    """List all non-soft-deleted tenants."""
    body = await _http.request_json("GET", "/admin/tenants")
    return TenantList.model_validate(body or {"tenants": []})


async def get(id: str) -> TenantDetail:  # noqa: A002 — positional ``id`` matches the dashboard API
    """Fetch a tenant's joined detail view (tenant + members + keys + usage)."""
    if not id:
        raise ValueError("tenant id must be a non-empty string")
    body = await _http.request_json("GET", f"/admin/tenants/{quote(id, safe='')}")
    return TenantDetail.model_validate(body or {})


async def create(input: CreateTenantInput | dict[str, Any]) -> Tenant:  # noqa: A002
    """Create a tenant. Accepts a :class:`CreateTenantInput` or a plain dict.

    The server validates the slug charset, length, and uniqueness. A
    duplicate slug raises :class:`AFStackError` with ``code == "CONFLICT"``.
    """
    payload = _as_dict(input, CreateTenantInput)
    body = await _http.request_json("POST", "/admin/tenants", json=payload)
    return Tenant.model_validate(body or {})


async def update(
    id: str,  # noqa: A002 — positional ``id`` matches the dashboard API
    input: UpdateTenantInput | dict[str, Any],
) -> Tenant:
    """Update selected tenant fields.

    Fields omitted from ``input`` are left untouched; passing an empty
    map for ``settings``/``quota`` clears them. The server returns the
    full updated row.
    """
    if not id:
        raise ValueError("tenant id must be a non-empty string")
    payload = _as_dict(input, UpdateTenantInput)
    body = await _http.request_json("PATCH", f"/admin/tenants/{quote(id, safe='')}", json=payload)
    return Tenant.model_validate(body or {})


async def delete(id: str) -> bool:  # noqa: A001, A002
    """Soft-delete a tenant. Returns ``True`` on success.

    Hard delete is intentionally not exposed; the runtime sets
    ``deleted_at = now()`` so audits remain queryable.
    """
    if not id:
        raise ValueError("tenant id must be a non-empty string")
    body = await _http.request_json("DELETE", f"/admin/tenants/{quote(id, safe='')}")
    if not isinstance(body, dict):
        return True
    return bool(body.get("deleted", True))


def _as_dict(
    value: Any,
    model: type[CreateTenantInput | UpdateTenantInput],
) -> dict[str, Any]:
    """Coerce ``value`` (dict OR ``BaseModel``) into a clean wire dict.

    Pydantic's ``model_dump(exclude_none=True)`` drops ``None`` fields so
    we don't accidentally null-out server state on partial updates.
    """
    if isinstance(value, model):
        return value.model_dump(exclude_none=True)
    if isinstance(value, dict):
        return {k: v for k, v in value.items() if v is not None}
    raise TypeError(f"input must be a {model.__name__} or dict, got {type(value).__name__}")


__all__ = ["create", "delete", "get", "list", "update"]
